subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c

```diff
@@ -224,53 +325,203 @@ static int validate_ak_order(void)
+{
+	return fp_settings_name_gen(name, SETTINGS_BOND_INFO_NAME_MAX_SIZE,
+				    SETTINGS_BOND_INFO_FULL_PREFIX, index);
+}
+
+static size_t fp_bond_info_idx_get(struct fp_bond_info *bond)
+{
+	return bond - fp_bonds;
+}
+
+static int fp_bond_info_settings_delete(struct fp_bond_info *bond)
 {
-	uint8_t id;
-	int first_zero_idx = -1;
 	int err;
+	char name[SETTINGS_BOND_INFO_NAME_MAX_SIZE];
 
-	if (settings_set_err) {
-		return settings_set_err;
+	err = fp_bond_info_name_gen(name, fp_bond_info_idx_get(bond));
+	if (err) {
+		return err;
 	}
 
-	for (size_t i = 0; i < ACCOUNT_KEY_CNT; i++) {
-		id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
-		if (id == 0) {
-			first_zero_idx = i;
-			break;
-		}
+	err = settings_delete(name);
+	if (err) {
+		return err;
+	}
 
-		if (account_key_id_to_idx(id) != i) {
+	return 0;
+}
+
+static int bond_remove(struct fp_bond_info *bond, bool bt_stack_bond_remove)
+{
+	int err;
+
+	__ASSERT_NO_MSG(!bt_addr_le_eq(&bond->nv.addr, BT_ADDR_LE_ANY));
+
+	if (bt_stack_bond_remove) {
+		if (!bt_request_cb) {
 			return -EINVAL;
 		}
+
+		err = bt_request_cb->bond_remove(&bond->nv.addr);
+		if (err) {
+			if (err == -ESRCH) {
+				LOG_DBG("Bond not found in the Bluetooth stack. "
+					"Continuing the removal operation");
+			} else {
+				LOG_ERR("Failed to remove Bluetooth bond (err %d)", err);
+				return err;
+			}
+		}
 	}
 
-	if (first_zero_idx != -1) {
-		for (size_t i = 0; i < first_zero_idx; i++) {
-			id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
-			if (id != ACCOUNT_KEY_MIN_ID + i) {
-				return -EINVAL;
+	/* Bond entry might have been deleted earlier by the fp_storage_ak_bond_delete. */
+	if (!bt_addr_le_eq(&bond->nv.addr, BT_ADDR_LE_ANY)) {
+		err = fp_bond_info_settings_delete(bond);
+		if (err) {
+			LOG_ERR("Failed to delete Fast Pair bond from Settings (err %d)", err);
+			return err;
+		}
+
+		*bond = (struct fp_bond_info){0};
+	}
+
+	return 0;
+}
+
+static void fp_bond_info_duplicates_handle(void)
+{
+	FP_BONDS_FOREACH(bond) {
+		bool duplicate_found = false;
+		bt_addr_le_t addr = bond->nv.addr;
+		int err;
+
+		if (bt_addr_le_eq(&bond->nv.addr, BT_ADDR_LE_ANY)) {
+			continue;
+		}
+
+		for (struct fp_bond_info *duplicate = bond + 1;
+		     duplicate <= &fp_bonds[ARRAY_SIZE(fp_bonds) - 1]; duplicate++) {
+			if (bt_addr_le_eq(&duplicate->nv.addr, &addr)) {
+				/* Error during Fast Pair Procedure. */
+				LOG_DBG("Removing duplicated bond entries at bootup");
+				if (!duplicate_found) {
+					duplicate_found = true;
+					err = bond_remove(duplicate, true);
+				} else {
+					err = bond_remove(duplicate, false);
+				}
+				if (err) {
+					LOG_ERR("bond_remove failed (err %d)", err);
+				}
 			}
 		}
 
-		for (size_t i = first_zero_idx + 1; i < ACCOUNT_KEY_CNT; i++) {
-			id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
-			if (id != 0) {
-				return -EINVAL;
+		/* The bond might have already been removed by fp_storage_ak_bond_delete. */
```

Comment:
The module that handles Bluetooth bond request callbacks calls fp_storage_ak_bond_delete to remove fp_bond_info when a Bluetooth bond is deleted for any reason. Because of this, the fp_bond_info might have already been cleared at this point.